### PR TITLE
Added unittests for snapshot method

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- The unittests for snapshot method is added.
- This checks if the minimum staking information is applied properly to compare the number of validators and demoted validators.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: consensus/istanbul/backend/backend_test.go [hunk 1]
```diff
@@ -865,11 +865,16 @@ func checkInCommitteeBlocks(seq int64, round int64) bool {
...
+}
+
+func newTestBackendWithConfig(chainConfig *params.ChainConfig) (b *backend) {
 	dbm := database.NewDBManager(&database.DBConfig{DBType: database.MemoryDB})
 	key, _ := crypto.GenerateKey()
 	istanbul.DefaultConfig.ProposerPolicy = istanbul.WeightedRandom
+	gov := governance.NewGovernanceInitialize(chainConfig, dbm)
 
-	backend := New(getTestRewards()[0], istanbul.DefaultConfig, key, dbm, getGovernance(dbm), common.CONSENSUSNODE).(*backend)
+	backend := New(getTestRewards()[0], istanbul.DefaultConfig, key, dbm, gov, common.CONSENSUSNODE).(*backend)
```
Comment (aidan-kwon) [inline consensus/istanbul/backend/backend_test.go:877]: `gov` is created with `chainConfig` containing an istanbul config, but `backend` is created with another istanbul config, `istanbul.DefaultConfig`. Could you unity those values to prevent mis-configuration between `gov` and `backend`?

Comment (jeongkyun-oh) [reply consensus/istanbul/backend/backend_test.go:877]: Hmm. I looked into it, but I think we do not use backend istanbul config.

Comment (jeongkyun-oh) [reply consensus/istanbul/backend/backend_test.go]: @aidan-kwon I have updated istanbulConfig to get from chainConfig. PTAL.

File: consensus/istanbul/backend/engine_test.go [hunk 2]
```diff
@@ -506,3 +577,126 @@ func TestWriteCommittedSeals(t *testing.T) {
...
+	}
+	return stakingInfo
+}
+
+func TestSnapshot(t *testing.T) {
+	type testcase struct {
+		stakingAmounts      []uint64 // test staking amounts of each validator
+		isIstanbulComatible bool     // whether or not if the inserted block is istanbul compatible
+		expectedValidators  int      // the number of expected validators
+		expectedDemoted     int      // the number of expected demoted validators
```
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:612]: What about `expectedValidatorNum` and `expectedDemotedNum`?

File: consensus/istanbul/backend/engine_test.go [hunk 3]
```diff
@@ -506,3 +577,126 @@ func TestWriteCommittedSeals(t *testing.T) {
...
+		stakingInfo.CouncilStakingAmounts = append(stakingInfo.CouncilStakingAmounts, amounts[idx])
+		stakingInfo.CouncilRewardAddrs = append(stakingInfo.CouncilRewardAddrs, rewardAddr)
+	}
+	return stakingInfo
+}
+
+func TestSnapshot(t *testing.T) {
+	type testcase struct {
+		stakingAmounts      []uint64 // test staking amounts of each validator
+		isIstanbulComatible bool     // whether or not if the inserted block is istanbul compatible
```
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:610]:
```suggestion
		isIstanbulCompatible bool     // whether or not if the inserted block is istanbul compatible
```

File: consensus/istanbul/backend/engine_test.go [hunk 4]
```diff
@@ -506,3 +577,126 @@ func TestWriteCommittedSeals(t *testing.T) {
...
+		expectedValidators  int      // the number of expected validators
+		expectedDemoted     int      // the number of expected demoted validators
+	}
+
+	testcases := []testcase{
+		// The following testcases are the ones before istanbul incompatible change
+		{
+			[]uint64{5000000, 5000000, 5000000, 5000000},
```
Comment (KimKyungup) [inline consensus/istanbul/backend/engine_test.go:618]: @jeongkyun-oh Could you add more edge cases like below?

The cases that the staking amount is ms-1, ms+1, ms, or 0.

If they are covered by other test cases, forget it. :)

**Review by KimKyungup — APPROVED** [review]

**Review by KimKyungup — DISMISSED** [review]
Thanks for the reflection.

**Review by yoomee1313 — DISMISSED** [review]

File: consensus/istanbul/backend/engine_test.go [hunk 5]
```diff
@@ -506,3 +577,170 @@ func TestWriteCommittedSeals(t *testing.T) {
...
+	configItems := makeSnapshotTestConfigItems()
+	configItems = append(configItems, minimumStake(new(big.Int).SetUint64(ms)))
+	for _, tc := range testcases {
+		if tc.isIstanbulCompatible {
+			configItems = append(configItems, istanbulCompatibleBlock(new(big.Int).SetUint64(0)))
+		}
+		chain, engine := newBlockChain(testNum, configItems...)
+
+		stakingInfo := makeFakeStakingInfo(0, nodeKeys, tc.stakingAmounts)
+		reward.SetTestStakingManagerWithStakingInfoCache(stakingInfo)
```
Comment (aidan-kwon) [inline consensus/istanbul/backend/engine_test.go:714]: `SetTestStakingManagerWithStakingInfoCache` changes a global variable, stakingManager. I am afraid that the changed value affects other tests executed later. How about changing back to the original value when the test is finished.

**Review by aidan-kwon — COMMENTED** [review]
LGTM except for a comment.

File: consensus/istanbul/backend/engine_test.go (hunk 5 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go:718]: @aidan-kwon Reflected :) PTAL.

File: consensus/istanbul/backend/engine_test.go [hunk 6]
```diff
@@ -506,3 +577,175 @@ func TestWriteCommittedSeals(t *testing.T) {
...
+			for idx, sa := range tc.stakingAmounts {
+				if sa >= ms {
+					assert.True(t, checkInValidators(addrs[idx], snap.ValSet.List()))
+				} else {
+					assert.True(t, checkInValidators(addrs[idx], snap.ValSet.DemotedList()))
+				}
+			}
+		} else {
+			// if the inserted block is not istanbul compatible
+			// and all validators doesn't have enough KLAYs
```
Comment (ehnuje) [inline consensus/istanbul/backend/engine_test.go:741]:
```suggestion
			// or all validators don't have enough KLAYs
```
Isn't it `or` if above `if` statement does not meet the condition?

File: consensus/istanbul/backend/engine_test.go [hunk 7]
```diff
@@ -506,3 +577,175 @@ func TestWriteCommittedSeals(t *testing.T) {
...
+	}
+	return true
+}
+
+func TestSnapshot(t *testing.T) {
+	type testcase struct {
+		stakingAmounts        []uint64 // test staking amounts of each validator
+		isIstanbulCompatible  bool     // whether or not if the inserted block is istanbul compatible
+		expectedValidatorsNum int      // the number of expected validators
+		expectedDemotedNum    int      // the number of expected demoted validators
```
Comment (ehnuje) [inline consensus/istanbul/backend/engine_test.go:629]: Not necessarily change this PR, however, please consider my comment in other PR. In my opinion, it is better to list up the indexes of expected validators/demoted, instead of their total numbers. I think direct comparing the expected outputs to the actual outputs is the purpose of the unittest.

**Review by ehnuje — APPROVED** [review]
LGTM except for minor comment. Thanks!

File: consensus/istanbul/backend/engine_test.go (hunk 6 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go]: You are right. Reflected :)

File: consensus/istanbul/backend/engine_test.go (hunk 7 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go:629]: Thanks for the comment. I will apply it on the next PR.

**Review by ehnuje — APPROVED** [review]

**Review by KimKyungup — APPROVED** [review]

**Review by yoomee1313 — APPROVED** [review]
